host/volume/api/http.go

```diff
@@ -0,0 +1,127 @@
+package volumeapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/julienschmidt/httprouter"
+	"github.com/flynn/flynn/host/volume"
+	"github.com/flynn/flynn/host/volume/zfs"
+	"github.com/flynn/flynn/pkg/httphelper"
+	"github.com/flynn/flynn/pkg/random"
+)
+
+type HTTPAPI struct {
+	vman *volume.Manager
+}
+
+func NewHTTPAPI(vman *volume.Manager) *HTTPAPI {
+	return &HTTPAPI{vman: vman}
+}
+
+func RegisterRoutes(api *HTTPAPI, r *httprouter.Router) {
+	r.POST("/storage/providers", api.CreateProvider)
+	r.POST("/storage/providers/:provider_id/volumes", api.Create)
+	r.GET("/storage/volumes", api.List)
+	r.GET("/storage/volumes/:volume_id", api.Inspect)
+	r.PUT("/storage/volumes/:volume_id/snapshot", api.Snapshot)
+}
+
+func (api *HTTPAPI) CreateProvider(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	var provider volume.Provider
+	pspec := &volume.ProviderSpec{}
+	if err := json.NewDecoder(r.Body).Decode(&pspec); err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+	switch pspec.Kind {
+	case "zfs":
+		config := &zfs.ProviderConfig{}
+		if err := json.Unmarshal(pspec.Config, config); err != nil {
+			httphelper.Error(w, err)
+			return
+		}
+		var err error
+		if provider, err = zfs.NewProvider(config); err != nil {
+			httphelper.Error(w, err)
+			return
+		}
+	case "":
+		httphelper.Error(w, httphelper.JSONError{
+			Code:    httphelper.ValidationError,
+			Message: fmt.Sprintf("volume provider 'kind' field must not be blank"),
+		})
+		return
+	default:
+		httphelper.Error(w, httphelper.JSONError{
+			Code:    httphelper.ValidationError,
+			Message: fmt.Sprintf("volume provider kind %q is not known"),
```

Comment:
Fixed instance; what's the invocation I should use to find more?  `go vet ./host/...` didn't emit any warnings for me.  Checked with `-x` and it was covering this (warn-worth) file.